### PR TITLE
feat(spanner): add samples for query with GFE, gRPC and query stats

### PR DESCRIPTION
**Pull request by songy23**
Go counterpart of https://github.com/GoogleCloudPlatform/java-docs-samples/tree/master/spanner/opencensus.

### CODE REVIEW

**Comment by songy23** [comment]
@mayurkale22

File: spanner/spanner_snippets/spanner/spanner_capture_grpc_metric.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanner
+
+// [START spanner_capture_grpc_metric]
```
Comment (mayurkale22) [inline spanner/spanner_snippets/spanner/spanner_capture_grpc_metric.go:17]:
```suggestion
// [START spanner_opencensus_capture_grpc_metric]
```

File: spanner/spanner_snippets/spanner/spanner_capture_grpc_metric.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		var singerID, albumID int64
+		var albumTitle string
+		if err := row.Columns(&singerID, &albumID, &albumTitle); err != nil {
+			return err
+		}
+		fmt.Fprintf(w, "%d %d %s\n", singerID, albumID, albumTitle)
+	}
+}
+
+// [END spanner_capture_grpc_metric]
```
Comment (mayurkale22) [inline spanner/spanner_snippets/spanner/spanner_capture_grpc_metric.go:79]:
```suggestion
// [END spanner_opencensus_capture_grpc_metric]
```

File: spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanner
+
+// [START spanner_capture_gfe_metric]
```
Comment (mayurkale22) [inline spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go:17]:
```suggestion
// [START spanner_opencensus_capture_gfe_metric]
```

File: spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	ctx, err = tag.New(ctx, tag.Insert(KeyMethod, "ExecuteSql"))
+	if err != nil {
+		return err
+	}
+	stats.Record(ctx, GFELatencyMs.M(int64(gfeLtcy)))
+
+	return nil
+}
+
+// [END spanner_capture_gfe_metric]
```
Comment (mayurkale22) [inline spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go:123]:
```suggestion
// [END spanner_opencensus_capture_gfe_metric]
```

File: spanner/spanner_snippets/spanner/spanner_capture_query_stats_metric.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanner
+
+// [START spanner_capture_query_stats_metric]
```
Comment (mayurkale22) [inline spanner/spanner_snippets/spanner/spanner_capture_query_stats_metric.go:17]:
```suggestion
// [START spanner_opencensus_capture_query_stats_metric]
```

File: spanner/spanner_snippets/spanner/spanner_capture_query_stats_metric.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+		var singerID, albumID int64
+		var albumTitle string
+		if err := row.Columns(&singerID, &albumID, &albumTitle); err != nil {
+			return err
+		}
+		fmt.Fprintf(w, "%d %d %s\n", singerID, albumID, albumTitle)
+	}
+}
+
+// [END spanner_capture_query_stats_metric]
```
Comment (mayurkale22) [inline spanner/spanner_snippets/spanner/spanner_capture_query_stats_metric.go:106]:
```suggestion
// [END spanner_opencensus_capture_query_stats_metric]
```

File: spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	session, err := client.CreateSession(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	// Execute a SQL query and retrieve the GFE server-timing header in gRPC metadata.
+	req2 := &sppb.ExecuteSqlRequest{
+		Session:   session.Name,
+		Sql:       `SELECT SingerId, AlbumId, AlbumTitle FROM Albums`,
+		QueryMode: sppb.ExecuteSqlRequest_PROFILE,
```
Comment (mayurkale22) [inline spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go:93]: I thought PROFILE must be set in query_stats sample and not here. I mean this is what I did in Java samples.

Comment (songy23) [reply spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go]: You're right it isn't needed in this sample. Removed.

File: spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go (hunk 3 above)
Comment (songy23) [reply spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go]: Done.

File: spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go (hunk 4 above)
Comment (songy23) [reply spanner/spanner_snippets/spanner/spanner_capture_gfe_metric.go]: Done.

File: spanner/spanner_snippets/spanner/spanner_capture_grpc_metric.go (hunk 1 above)
Comment (songy23) [reply spanner/spanner_snippets/spanner/spanner_capture_grpc_metric.go]: Done.

File: spanner/spanner_snippets/spanner/spanner_capture_grpc_metric.go (hunk 2 above)
Comment (songy23) [reply spanner/spanner_snippets/spanner/spanner_capture_grpc_metric.go]: Done.

File: spanner/spanner_snippets/spanner/spanner_capture_query_stats_metric.go (hunk 5 above)
Comment (songy23) [reply spanner/spanner_snippets/spanner/spanner_capture_query_stats_metric.go]: Done.

File: spanner/spanner_snippets/spanner/spanner_capture_query_stats_metric.go (hunk 6 above)
Comment (songy23) [reply spanner/spanner_snippets/spanner/spanner_capture_query_stats_metric.go]: Done.

**Review by mayurkale22 — APPROVED** [review]
Looks good.

File: spanner/spanner_snippets/spanner/integration_test.go [hunk 8]
```diff
@@ -45,6 +45,7 @@ type sampleFuncWithContext func(ctx context.Context, w io.Writer, dbName string)
 type instanceSampleFunc func(w io.Writer, projectID, instanceID string) error
 type backupSampleFunc func(ctx context.Context, w io.Writer, dbName, backupID string) error
 type createBackupSampleFunc func(ctx context.Context, w io.Writer, dbName, backupID string, versionTime time.Time) error
+type sampleFuncWithProjectID func(w io.Writer, dbName, projectID string) error
```
Comment (hengfengli) [inline spanner/spanner_snippets/spanner/integration_test.go:48]: Can we reuse the existing `sampleFunc` and extract `projectID` from `dbName` in the samples? If not, can we change the order of the parameters to `projectID, dbName string`? 

I still think extracting `projectID` from `dbName` might be better, like: 

https://github.com/GoogleCloudPlatform/golang-samples/blob/296e1434dc002ab65f118c40dac5580dc002d451/spanner/spanner_snippets/spanner/integration_test.go#L713-L718

But of course, there are some code duplications. 

Or we should have a 

```go
type databaseSampleFunc func(w io.Writer, projectID, instanceID, databaseID string) error
```

Comment (songy23) [reply spanner/spanner_snippets/spanner/integration_test.go]: Makes sense, changed to parse db name to get project IDs.

**Review by hengfengli — APPROVED** [review]
Thanks for the changes. LGTM.

**Pull request merged**